### PR TITLE
add stream lock to avoid use same steam in multithread at same time

### PR DESCRIPTION
**Pull request by zhaoguochun1995**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation and Context

1. 背景： 多线程场景下，不支持调用aclopCompileAndExecute接口时指定同一个Stream或使用默认Stream，否则可能任务执行异常。（https://www.hiascend.com/document/detail/zh/canncommercial/70RC1/reference/envvar/envref_07_0047.html?sub_id=%2Fzh%2Fcanncommercial%2F70RC1%2Finferapplicationdev%2Faclcppdevg%2Faclcppdevg_03_0632.html）
2. 调用aclopCompileAndExecute前为同一个stream加锁

## Checklist
**Before PR**:

- [ ] I have read and followed the workflow indicated in the [Contributors.md](https://github.com/DeepLink-org/DIOPI/blob/main/Contributors.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [Contributors.md](https://github.com/DeepLink-org/DIOPI/blob/main/Contributors.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Review by hellozmz — APPROVED** [review]

**Review by jfxu-st — APPROVED** [review]

File: impl/ascend/common/stream_lock.cpp [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+#include <thread>
+
+namespace diopi {
+
+using MutexType = std::mutex;
+
+namespace {
+
+MutexType* getLockForStream(void* aclrtStreamHandle) {
+    static std::map<void*, MutexType> streamThreadMutexMap;
```
Comment (hellozmz) [inline impl/ascend/common/stream_lock.cpp:14]: 以后程序运行起来，这个map可能比较大，有内存泄漏的风险。

整体设计的还挺巧妙的👍

Comment (zhaoguochun1995) [reply impl/ascend/common/stream_lock.cpp:14]: map 的size是 dipu中steam的数量，不会很多的。

Comment (jfxu-st) [reply impl/ascend/common/stream_lock.cpp:14]: 为啥要把 getLockForStream 函数以及 streamThreadMutexMap 变量单独拿出来？放在 StreamLockGuard 类里面，作为静态成员变量，会更合适吗？

Comment (zhaoguochun1995) [reply impl/ascend/common/stream_lock.cpp:14]: 1. streamThreadMutexMap的生命周期不一样
2. 头文件里不想有太多的依赖
3. 接口和实现解耦

**Review by yangbofun — APPROVED** [review]

**Pull request merged**